src/ios/CDVLocation.m

```diff
@@ -169,199 +180,221 @@ - (void)_stopLocation
-                    }
-                    // Tell the location manager to start notifying us of heading updates
-                    [strongSelf startLocation:enableHighAccuracy];
-                } else {
-                    [strongSelf returnLocationInfo:callbackId andKeepCallback:NO];
+            [strongSelf prepareLocationData];
+            
+            if (!strongSelf->__locationStarted || strongSelf->__highAccuracyEnabled != enableHighAccuracy) {
+                if (callbackId) {
+                    [strongSelf.locationData.locationCallbacks addObject:callbackId];
                 }
+                [strongSelf startLocation:enableHighAccuracy];
+            } else {
+                [strongSelf returnLocationInfo:callbackId andKeepCallback:NO];
             }
         }];
     }];
 }
 
-- (void)addWatch:(CDVInvokedUrlCommand*)command
+- (void)addWatch:(CDVInvokedUrlCommand *)command
 {
     __weak __typeof(self) weakSelf = self;
-    NSString* callbackId = command.callbackId;
-    NSString* timerId = [command argumentAtIndex:0];
+    NSString *callbackId = command.callbackId;
+    NSString *timerId = [command argumentAtIndex:0];
     BOOL enableHighAccuracy = [[command argumentAtIndex:1] boolValue];
 
-    if (!self.locationData) {
-        self.locationData = [[CDVLocationData alloc] init];
-    }
-    CDVLocationData* lData = self.locationData;
-
-    if (!lData.watchCallbacks) {
-        lData.watchCallbacks = [NSMutableDictionary dictionaryWithCapacity:1];
-    }
+    [self prepareLocationData];
 
-    // add the callbackId into the dictionary so we can call back whenever get data
-    [lData.watchCallbacks setObject:callbackId forKey:timerId];
+    self.locationData.watchCallbacks[timerId] = callbackId;
 
     [self isLocationServicesEnabledWithCompletion:^(BOOL enabled) {
         __strong __typeof(weakSelf) strongSelf = weakSelf;
         if (!strongSelf) {
             return;
         }
-        
-        if (enabled == NO) {
+
+        if (!enabled) {
             [strongSelf returnLocationError:PERMISSIONDENIED withMessage:@"Location services are disabled."];
-        } else if (!strongSelf->__locationStarted || (strongSelf->__highAccuracyEnabled != enableHighAccuracy)) {
-            // Tell the location manager to start notifying us of location updates
+            return;
+        }
+
+        if (!strongSelf->__locationStarted || strongSelf->__highAccuracyEnabled != enableHighAccuracy) {
             [strongSelf startLocation:enableHighAccuracy];
         }
     }];
 }
 
-- (void)clearWatch:(CDVInvokedUrlCommand*)command
+- (void)clearWatch:(CDVInvokedUrlCommand *)command
 {
-    NSString* timerId = [command argumentAtIndex:0];
+    NSString *timerId = [command argumentAtIndex:0];
 
-    if (self.locationData && self.locationData.watchCallbacks && [self.locationData.watchCallbacks objectForKey:timerId]) {
+    if (self.locationData && self.locationData.watchCallbacks[timerId]) {
```

Comment:
You removed the `self.locationData.watchCallbacks` verification. Was this on purpose?

Reply:
Yes, because it doesn't really matter when we have inside the if this check: `self.locationData.watchCallbacks[timerId]`

Reply:
It's the other way around: if you do `self.locationData.watchCallbacks[timerId]` and `self.locationData.watchCallbacks` is NULL, it will crash on runtime.
The crash will not happen if you use `[self.locationData.watchCallbacks objectForKey:timerId]`, so the old approach, although not perfect, was better than the new one.

Reply:
Fixed